File: frontend/src/services/citationService.js
import api from './api';

const citationService = {
    // Generate APA citation
    generateAPACitation: async (paperId) => {
        try {
            const response = await api.get(`/citations/generate/apa/${paperId}`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Generate IEEE citation
    generateIEEECitation: async (paperId) => {
        try {
            const response = await api.get(`/citations/generate/ieee/${paperId}`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Generate Chicago citation
    generateChicagoCitation: async (paperId) => {
        try {
            const response = await api.get(`/citations/generate/chicago/${paperId}`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Generate MLA citation
    generateMLACitation: async (paperId) => {
        try {
            const response = await api.get(`/citations/generate/mla/${paperId}`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Save citation
    saveCitation: async (citationData) => {
        try {
            const response = await api.post('/citations/save', citationData);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Get user citations
    getUserCitations: async (userId) => {
        try {
            const response = await api.get(`/citations/user/${userId}`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Delete citation
    deleteCitation: async (citationId) => {
        try {
            const response = await api.delete(`/citations/${citationId}`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },
};

export default citationService;

File: src/main/java/com/researchpaper/service/CitationService.java
package com.researchpaper.service;

import com.researchpaper.entity.Citation;
import com.researchpaper.entity.ResearchPaper;
import com.researchpaper.entity.User;
import com.researchpaper.repository.CitationRepository;
import com.researchpaper.repository.ResearchPaperRepository;
import com.researchpaper.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.List;

@Service
public class CitationService {

    @Autowired
    private CitationRepository citationRepository;

    @Autowired
    private ResearchPaperRepository paperRepository;

    @Autowired
    private UserRepository userRepository;

    // Generate APA Citation Format
    public String generateAPACitation(ResearchPaper paper) {
        return String.format("%s (%d). %s. Retrieved from %s",
                paper.getAuthors(),
                paper.getPublicationYear(),
                paper.getTitle(),
                paper.getUrl());
    }

    // Generate IEEE Citation Format
    public String generateIEEECitation(ResearchPaper paper) {
        return String.format("[1] %s, \"%s,\" %d. [Online]. Available: %s",
                paper.getAuthors(),
                paper.getTitle(),
                paper.getPublicationYear(),
                paper.getUrl());
    }

    // Generate Chicago Citation Format
    public String generateChicagoCitation(ResearchPaper paper) {
        return String.format("%s. \"%s.\" %d. Accessed from %s",
                paper.getAuthors(),
                paper.getTitle(),
                paper.getPublicationYear(),
                paper.getUrl());
    }

    // Generate MLA Citation Format
    public String generateMLACitation(ResearchPaper paper) {
        return String.format("%s. \"%s.\" %d. Web. %s",
                paper.getAuthors(),
                paper.getTitle(),
                paper.getPublicationYear(),
                paper.getUrl());
    }

    // Save citation with paper and user relationships
    public Citation saveCitation(Citation citation) {
        citation.setCreatedAt(LocalDateTime.now());
        return citationRepository.save(citation);
    }

    // Get citations by user ID
    public List<Citation> getCitationsByUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));
        return citationRepository.findByUser(user);
    }

    // Get citations by paper ID
    public List<Citation> getCitationsByPaper(Long paperId) {
        ResearchPaper paper = paperRepository.findById(paperId)
                .orElseThrow(() -> new IllegalArgumentException("Paper not found"));
        return citationRepository.findByPaper(paper);
    }

    // Delete citation by ID
    public void deleteCitation(Long citationId) {
        citationRepository.deleteById(citationId);
    }

    // Get citation by ID
    public Citation getCitationById(Long citationId) {
        return citationRepository.findById(citationId)
                .orElseThrow(() -> new IllegalArgumentException("Citation not found"));
    }
}